Create minimum amount of AMP processors

michalczaplinski opened the pull request:
**What**:

Create minimum amount of AMP processors.

Add e2e tests that validate that the HTML is valid AMP markup after being processed by AMP processors.

**How**:

Add the processors to the AMP package.


**Tasks**:

- [x] Code
- [x] TSDocs
- [x] TypeScript
- [x] Unit tests
- [x] End to end tests
- [x] Add a changeset (with link to its [Feature Discussion](https://community.frontity.org/c/33) if it exists)

**Unrelated Tasks**

<!-- ignore-task-list-start -->
- [ ] Update community discussions
- [ ] TypeScript tests
- [ ] Update starter themes
- [ ] Update other packages
- [ ] Open an issue for this feature in the [docs repo](https://github.com/frontity/docs/wiki/Code-Releases)
<!-- ignore-task-list-end -->

michalczaplinski committed:
Add the boilerplate for adding AMP processors

michalczaplinski committed:
Add e2e tests for amp package

michalczaplinski committed:
Validate that the content is valid AMP

michalczaplinski committed:
Remove the test.html file that should not be there

michalczaplinski committed:
Fixes for tests for AMP, removing the uploads

michalczaplinski committed:
Merge branch 'dev' of github.com:frontity/frontity into amp-processors

michalczaplinski committed:
Remove the parts of the amp e2e test that were not working

michalczaplinski committed:
Fix the amp.test.ts adding the html2react package

michalczaplinski committed:
First basic amp test

michalczaplinski committed:
Move img processor to folder and improve amp processor types

michalczaplinski committed:
Rename image processor to img

michalczaplinski committed:
Remove react-testing-library and add jest-amp

michalczaplinski committed:
Fix iframe processor and add proper tests

michalczaplinski committed:
Add width and height to amp-image and amp-iframe

michalczaplinski committed:
Merge branch 'dev' of github.com:frontity/frontity into amp-processors

michalczaplinski committed:
Update package-lock.json

michalczaplinski commented:
Update on current status of this PR: 

https://www.loom.com/share/7a167b830ede49b29ef333e66ad4fa67

michalczaplinski committed:
Replace async=true with async in the processor tests

michalczaplinski committed:
Add video processor

michalczaplinski committed:
Add processors for audio, removing tags & classes

michalczaplinski committed:
Add processors to remove IDs

michalczaplinski committed:
Adding the twitter processor

michalczaplinski committed:
Merge branch 'dev' of github.com:frontity/frontity into amp-processors

michalczaplinski committed:
Add the processors import again

michalczaplinski committed:
Update package-lock.json

michalczaplinski committed:
Fix frontity.settings.ts in e2e tests

michalczaplinski committed:
Add an eslint override fixing eslint issue in cypress expect

michalczaplinski committed:
Remove removing scripts until we merge the PR to expose head & script.

michalczaplinski committed:
Create separate amp-wordpress package in frontity.settings.ts

michalczaplinski committed:
Merge branch 'dev' of github.com:frontity/frontity into amp-processors

michalczaplinski committed:
Update package-lock.json

michalczaplinski committed:
Remove unnecesary types from e2e/amp

michalczaplinski committed:
Add names to processors

michalczaplinski committed:
Correct img AMP processor and fix its test

michalczaplinski committed:
Fix the amp audio processor

michalczaplinski committed:
More small fixes for amp processors

michalczaplinski committed:
Load amp iframe over HTTPS

michalczaplinski committed:
refactor amp-audio

michalczaplinski committed:
Multiple adjustments for AMP video and audio processors

michalczaplinski committed:
Alphabetize & dedupe deps in package.jsons

michalczaplinski committed:
Add twitter processor

michalczaplinski committed:
Add the youtube processor.

michalczaplinski committed:
Fix iframe processor

michalczaplinski committed:
Add container to amp e2e tests

michalczaplinski committed:
Add some more fixes and comments

michalczaplinski committed:
Do not add controls on amp-video

michalczaplinski committed:
Remove the controls prop from amp-video

michalczaplinski committed:
Remove unused rule in eslintrc

michalczaplinski committed:
Add additional comments for types

michalczaplinski committed:
Add warning if html2react is not installed in amp

michalczaplinski committed:
Fix amp unit tests

michalczaplinski committed:
Uncomment the after() in amp e2e tests

michalczaplinski committed:
Add eslint rule for the cypress expect

michalczaplinski committed:
Add comments to img processor and tests

michalczaplinski committed:
Add changeset

michalczaplinski committed:
Merge branches 'amp-processors' and 'dev' of github.com:frontity/frontity into amp-processors

michalczaplinski committed:
Update dependencies

michalczaplinski commented:
I've created a video with the summary of the changes: 

https://www.loom.com/share/ae2bc9dababa4f418819b1a9c2b086cd

One additional thing that I have not mentioned in the video is that I didn't create a soundcloud processor because I think that for the first version it's perhaps not necessary as soundcloud is probably used an order of magnitude less than other platforms for which we have created the processors now like youtube or twitter.

michalczaplinski committed:
Update dependencies

michalczaplinski committed:
Fix types in html2react

luisherranz commented on .changeset/strong-planets-wash.md:
```diff
@@ -0,0 +1,5 @@
+---
+"@frontity/amp": patch
```
```suggestion
"@frontity/amp": minor
```

I'd make this a minor.

luisherranz commented on .changeset/strong-planets-wash.md:
```diff
@@ -0,0 +1,5 @@
+"@frontity/amp": patch
+---
+
+Add first the version of html2react processors that are necessary for the AMP package.
```
Maybe the list of processors would be nice as well.

michalczaplinski replied:
ok, will add that! :)

luisherranz committed:
Use expect from Jest

luisherranz committed:
Add `have.attr` to some e2e tests

luisherranz commented on packages/amp/amp.d.ts:
```diff
@@ -0,0 +1,59 @@
+  /**
+   * Override the built in JSX interface.
+   */
+  interface IntrinsicElements {
```
@DAreRodz you have also added one of these in your PR. 

Do you know if there is a way to declare them together, or each unique package must declare its own dependencies?

DAreRodz replied:
Good question. We could use a workaround to add all AMP elements in the `@frontity/amp` package, something like:

```ts
declare global {
  namespace JSX {
    interface IntrinsicElements {
      [key: string]: any;
    }
  }
}
```

It would work because the types for the already-defined HTML elements are preserved.

In the [NextJS docs](https://nextjs.org/docs/advanced-features/amp-support/typescript), there's a link to a StackOverflow thread where that workaround is suggested. Also, they say that AMP has in their roadmap to add types for their elements, but it seems that's not going to happen in the near future (https://github.com/ampproject/amphtml/issues/13791#issuecomment-650460822).

luisherranz replied:
We can try that. See if it solves the problems on all packages.

@michalczaplinski what do you think? Do you agree? If so, could you refactor that?

michalczaplinski replied:
hmm, I'm not sure if adding 

```
interface IntrinsicElements {
  ...
  [key: string]: any;
}
```

would make sense here because then any string can be used as the JSX element and that makes some typos more likely. 

But more importantly, the `amp.d.ts` file is scoped to the AMP package so it would not solve the issue for all packages, only for the code that is in the AMP package.

If any other package would want to use that workaround, they would have to reference the types of AMP package like:

```
/// <reference types="@frontity/amp" />
```

This is because we are augumenting a namespace in the global scope and TS prevents it from leaking by default. It's the same reason why we have to e.g. reference the Cypress types in https://github.com/frontity/frontity/blob/ca55fa5d235ba297ed88f355459d63bcd2790f90/e2e/support/index.d.ts#L3-L3

So, in my opinion the best way would be to leave those types as they are, what do you think?

luisherranz replied:
Whatever works. If it can't be done or doesn't make sense, don't worry. It's just a matter of keeping everything as tidy and organized as possible 🙂

michalczaplinski replied:
ok, will keep it as it is in that case :)

luisherranz commented on packages/amp/src/__tests__/processors.test.tsx:
```diff
@@ -0,0 +1,219 @@
+HelmetProvider.canUseDOM = false;
+
+const replaceHeadAttributes = (head: HelmetData) =>
+  head.script.toString().replace(/async=("|')?true("|')?/, "async");
```
The AMP validator is what Google runs to validate if an AMP page is valid or not, so unfortunately we have to use `async=""`.

@darerodz also found this problem on his PR.

```suggestion
const replaceHeadAttributes = (head: HelmetData) => head.script.toString();
```

But it seems like your code is already fine with the `amp={undefined}` trick, because if I remove that replace the validator passes fine 🙂

luisherranz commented on packages/amp/src/index.tsx:
```diff
@@ -25,6 +26,15 @@ export default {
   actions: {
     amp: {
       init: ({ libraries }) => {
+        if (!libraries.html2react) {
```
The existence of `libraries.html2react` is not a guarantee that the `@html2react` package is installed because if another package adds processors like this:

```js
export default {
  libraries: {
    html2react: {
      processors: [
        /* my processors */
      ],
    },
  },
};
```

then `libraries.html2react` will be defined.

But don't worry just, let's leave it like this because we haven't figured out how to do this yet.

---

We were actually talking about this for the AMP package in the AMP FD, I don't know if you saw it: https://community.frontity.org/t/amp-package/388/49?u=luisherranz

michalczaplinski replied:
oh, that's a good point! thanks 😊

luisherranz commented on packages/amp/src/index.tsx:
```diff
@@ -25,6 +26,15 @@ export default {
+          );
+        }
+        // Add the AMP processors
+        libraries?.html2react?.processors?.push(...processors);
```
I think we can move this to the package definition, to be consistent with the rest of the processor definitions in other packages, don't you think?

michalczaplinski replied:
Totally right 👍

luisherranz commented on packages/amp/src/processors/audio.tsx:
```diff
@@ -0,0 +1,42 @@
+    // AMP requires that the file is loaded over HTTPS
+    const httpRegexp = /^http:\/\//;
+    if (node.props.src.match(httpRegexp)) {
+      node.props.src = node.props.src.replace(httpRegexp, "https://");
```
Maybe apart from the fix, we should emit another warning here to let the user know that this URL is not valid? 

I say so because maybe the `https` version of the URL doesn't exist and then the audio won't work and the user won't know why.

michalczaplinski replied:
you're right, I agree that we should warn about this 👍 

Just in case you were wondering, HTTPS is not needed for images so I didn't add that regex in the `amp-img` processor.

luisherranz commented on packages/amp/src/processors/iframe.tsx:
```diff
@@ -0,0 +1,70 @@
+    // AMP requires that the iframe is loaded over HTTPS
+    const httpRegexp = /^http:\/\//;
+    if (node.props.src.match(httpRegexp)) {
+      node.props.src = node.props.src.replace(httpRegexp, "https://");
```
Same here.

luisherranz commented on packages/amp/src/processors/removeProhibitedTags.ts:
```diff
@@ -0,0 +1,19 @@
+    node.type === "element" &&
+    (node.component === "style" ||
+      node.component === "script" ||
+      node.component === "picture" ||
```
Do you know if for `picture` we should return the children `img` tag to honor the fallback or that is not necessary?

Docs: https://www.w3schools.com/tags/tag_picture.asp

```html
<picture>
  <source media="(min-width:650px)" srcset="img_pink_flowers.jpg" />
  <source media="(min-width:465px)" srcset="img_white_flower.jpg" />
  <img src="img_orange_flowers.jpg" alt="Flowers" style="width:auto;" />
</picture>
```

michalczaplinski replied:
I think we could do that. I would however also throw a warning saying that this is happening so that the user could switch from `<picture>` to `<img>` in their code.

michalczaplinski replied:
While working on this, I've found this bug: https://github.com/frontity/frontity/issues/742 

It's not strictly a blocker I don't think we should merge until it's solved because if there is a `style` attribute on `img`,`iframe`, `audio`, etc. the AMP markup will not be correct.

luisherranz replied:
Good catch. I answered in the issue 🙂

luisherranz commented on packages/amp/src/processors/img.tsx:
```diff
@@ -0,0 +1,51 @@
+    // We can now safely remove the className prop.
+    delete node.props.className;
+
+    // amp-img does not support nor need that prop.
+    delete node.props.loading;
```
Is this true only for `amp-img` or also for `amp-iframe`? If so, should we add it to the `iframe` processor as well?

michalczaplinski replied:
Yes, the amp-iframe also does not allow the `loading` attribute, but the difference here is that WordPress will automatically add the `loading="lazy"` attribute on the images so we have to strip that. 

However the when creating an iframe the user would have to opt-in to using the `loading="lazy"` in their content in which case the AMP validator which runs in development in all AMP sites would fail informing them that this property is not allowed.

So, I think we can't go wrong either way, but maybe it would be slightly better to NOT remove the `loading` and just let the user fix their content. What do you think?

luisherranz replied:
> the difference here is that WordPress will automatically add the loading="lazy" attribute on the images

Yeah, but that's the current version. Maybe they also add native lazy loading to iframes in the next version. Or in the iframe Gutenberg block. Or maybe users use a plugin for that, like https://wordpress.org/plugins/lazy-loading-responsive-images/...

> maybe it would be slightly better to NOT remove the loading and just let the user fix their content. What do you think?

The problem is that the `post.content` is shared between the React version and the AMP version. If they remove it for AMP, they would be removing it from React as well.

michalczaplinski replied:
> The problem is that the post.content is shared between the React version and the AMP version. If they remove it for AMP, they would be removing it from React as well.


That's a fair point! Will remove the `loading` in the processor in that case.

luisherranz replied:
Awesome. Thanks Michal!

luisherranz reviewed (changes requested):
Great work! The AMP validation tools for Jest and Cypress are awesome 😄

> I didn't create a soundcloud processor

Sounds reasonable to me 👍

I've made a couple of small changes while I was reviewing the code:

- I've changed the Chai expect for Jest expect (we are using always the Jest's one, even in Cypress).
- I've added some `have.attr` to the e2e tests to make them slightly more robust.

---

I think that the only thing that is missing is the processor priorities. Right now they are in conflict with other processors, like the ones defined in `@frontity/html2react`.

What about using a priority of `9`? That way they will have precedence over the default priority processors (`priority: 10`).

Also, I wonder if we should have two separate `iframe` and `image` processors or make the `iframe` and `image` processors of `@frontity/html2react` AMP-aware.

What do you think?

cc: @santosguillamot

michalczaplinski committed:
Add more details to the changeset

michalczaplinski committed:
Remove line that replaces "async=true with "async" in amp package tests

Co-authored-by: Luis Herranz <luisherranz@gmail.com>

michalczaplinski committed:
Merge branch 'amp-processors' of github.com:frontity/frontity into amp-processors

michalczaplinski committed:
Remove eslint disable

michalczaplinski committed:
Add processors in the AMP package definition

michalczaplinski committed:
Add warning when http -> https in AMP processors

michalczaplinski committed:
Fix old import from @emotion/core to @emotion/react

luisherranz commented:
I have just realized that maybe we need to be a bit more careful with the warnings because the same `post.content` is going to be shared between the React app and the AMP pages.

That means that, for example, we could be throwing a warning to ask the user to replace a `<picture>` tag because that is not valid in AMP HTML. But that is valid in the React app and maybe that `<picture>` tag is exactly what they want for the React app.

So maybe we should not throw warnings for things that are perfectly valid in the React app. We should simply try to gracefully convert them to valid AMP HTML.

What do you think?

michalczaplinski committed:
Merge branch 'dev' of github.com:frontity/frontity into amp-processors

michalczaplinski committed:
Update package-lock.json

michalczaplinski committed:
Remove unnecessary node.type === element checks

michalczaplinski committed:
Add processor for picture element

michalczaplinski committed:
Add missing names to processors

michalczaplinski committed:
AMP: Update picture processor

michalczaplinski committed:
AMP: Remove warnings

michalczaplinski committed:
Merge branch 'dev' of github.com:frontity/frontity into amp-processors

michalczaplinski committed:
AMP: Pass processors without the array

michalczaplinski committed:
Remove resetting of database after e2e amp tests

michalczaplinski committed:
Add missing names to processors in html2react

michalczaplinski committed:
Add correct priorities to amp processors

michalczaplinski committed:
Update types of e2e/amp, add html2react processors

michalczaplinski committed:
Reset database after amp e2e tests

michalczaplinski committed:
Empty commit

michalczaplinski committed:
Add debug flag and catch err in install plugin command

michalczaplinski commented:
Hey @luisherranz 

I've recorded another summary for you: https://www.loom.com/share/960082d6e08e49f6a0ca44a085c54bf8

michalczaplinski committed:
Update cypress

michalczaplinski commented on .github/workflows/e2e-tests-chrome.yml:
```diff
@@ -47,7 +50,7 @@ jobs:
 
       - name: Run e2e tests
-        uses: cypress-io/github-action@v2
+        uses: cypress-io/github-action@v2.9.1
```
While investigating e2e test failures, I also noticed that we're not pinning the version of the GH action. I think we should be doing that just in case there is some subtle change.

luisherranz commented on packages/amp/src/index.tsx:
```diff
@@ -118,6 +126,11 @@ const amp: AMP = {
+      processors,
+    },
+  },
+} as AMP;
```
I don't think this is needed, right?

```suggestion
};
```

michalczaplinski replied:
nope, not needed. 👍

luisherranz commented on packages/amp/src/processors/img.tsx:
```diff
@@ -0,0 +1,52 @@
+    // Because amp-* component are custom components, we have to pass the
+    // class as `class` and not as `className`:
+    // https://reactjs.org/docs/web-components.html#using-web-components-in-react
+    node.props.class = node.props?.className;
```
Is this still needed after we merged this PR https://github.com/frontity/frontity/pull/741?

michalczaplinski replied:
You're right we don't need it anymore!

luisherranz commented on packages/amp/src/processors/picture.tsx:
```diff
@@ -0,0 +1,24 @@
+  name: "amp: replace picture with amp-img",
+  test: ({ node }) =>
+    node.component === "picture" &&
+    Boolean(node.children.find((node: Element) => node.component === "img")),
```
What happens with `<picture>` elements that don't have a `<img>` child? Are those allowed in AMP?

Also, should this have a similar priority to the `image` one, or is 10 (default) ok?

michalczaplinski replied:
`<picture>` is not allowed at all in AMP and `amp-image` should provide the the same functionality: https://amp.dev/documentation/guides-and-tutorials/learn/spec/amphtml/#html-tags

Also `<picture>` without an `<img>` is [not allowed in HTML](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/picture) so I don't think we need to handle it in any special way.

Good point about the priorities though, I've changed it to 8: 

https://github.com/frontity/frontity/blob/1869341ea2b847a0d58be9c3a76c179349c6cbbf/packages/amp/src/processors/picture.tsx#L8

and added an e2e for the `picture` element: 

https://github.com/frontity/frontity/blob/1869341ea2b847a0d58be9c3a76c179349c6cbbf/e2e/integration/wordpress-01/amp.spec.ts#L98-L107

luisherranz commented on packages/amp/src/processors/removeProhibitedTags.ts:
```diff
@@ -0,0 +1,17 @@
+import { Packages } from "../../types";
+
+export const removeProhibitedTags: Processor<Element, Packages> = {
+  name: "amp: removeProhibitedTags",
```
I guess this processor should have a rather low priority in case other processors are adding new nodes with these tags, shouldn't it?

What about `priority: 20`? 
```suggestion
  name: "amp: removeProhibitedTags",
  priority: 20,
```

luisherranz reviewed (approved):
I have a couple of comments which I added below, but I am going to approve anyway so feel free to merge it whenever you want.

Great work Michal! 🙂👍

michalczaplinski committed:
Merge branch 'dev' of github.com:frontity/frontity into amp-processors

michalczaplinski committed:
Update priority for removeProhibitedTags processor

Co-authored-by: Luis Herranz <luisherranz@gmail.com>

michalczaplinski committed:
Remove the "as AMP"

michalczaplinski committed:
Remove unnecessary code in amp-img processor

michalczaplinski committed:
Update priority of picture processor and add e2e test

michalczaplinski committed:
Remove the after() from the AMP e2e tests

michalczaplinski commented:
> I have a couple of comments which I added below, but I am going to approve anyway so feel free to merge it whenever you want.

Great, thanks Luis! I've added minor changes according to your comments and also I've removed out the `after()` in the AMP e2e tests because:

1. I think that we don't need to use it and shouldn't use it because all tests should take care of their own setup.

2. When developing locally, most of the time it's necessary to do some changes in the WP admin, run `node export-db.js` script and re-run the tests. For this to work we need to keep the DB state after the tests finish running :)
